### PR TITLE
Loss functions XGB

### PR DESCRIPTION
**Pull request by TirelessClock**
Started working on new loss functions more specific to XGB.
Continuing from PR #3747

### CODE REVIEW

File: src/mlpack/methods/decision_tree/log_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_LOG_LOSS_HPP
+#define MLPACK_METHODS_XGBOOST_LOG_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <cmath>
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/log_loss.hpp:16]: Trivial comment but I think that this is already included by `prereqs.hpp` or `base.hpp`. :+1:

File: src/mlpack/methods/decision_tree/log_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    // Return the log-odds of the mean of the values.
+    double mean = accu(values) / (typename VecType::elem_type) values.n_elem;
+    return std::log(mean / (1 - mean));
+  }
+
+  arma::rowvec Softmax(const arma::rowvec& logits) {
+      arma::rowvec exp_logits = arma::exp(logits - arma::max(logits));
+      return exp_logits / arma::accu(exp_logits);
+  }
+
```
Comment (shrit) [inline src/mlpack/methods/decision_tree/log_loss.hpp:57]:
```suggestion
```

File: src/mlpack/methods/decision_tree/log_loss.hpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+   * @param begin The begin index to calculate gain.
+   * @param end The end index to calculate gain.
+   */
+  template<bool UseWeights, typename MatType, typename WeightVecType>
+  double Evaluate(const MatType& input, 
+                  const WeightVecType& /* weights */,
+                  const size_t begin, 
+                  const size_t end)
+  {
+
```
Comment (shrit) [inline src/mlpack/methods/decision_tree/log_loss.hpp:71]:
```suggestion
```

File: src/mlpack/methods/decision_tree/log_loss.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  template<bool UseWeights, typename VecType, typename WeightVecType>
+  static double Evaluate(const VecType& values,
+                         const WeightVecType& /* weights */)
+  {
+    return Evaluate(values, 0, values.n_cols);
+  }
+
+
```
Comment (shrit) [inline src/mlpack/methods/decision_tree/log_loss.hpp:100]:
```suggestion
```

File: src/mlpack/methods/decision_tree/log_loss.hpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  {
+    if (sumGradients > alpha)
+    {
+      return sumGradients - alpha;
+    }
+    else if (sumGradients < - alpha)
+    {
+      return sumGradients + alpha;
+    }
+
```
Comment (shrit) [inline src/mlpack/methods/decision_tree/log_loss.hpp:123]:
```suggestion
```

**Review by shrit — CHANGES_REQUESTED** [review]
@TirelessClock could you add a couple of tests for this loss function ?
